mat/tridiag.go

```diff
@@ -0,0 +1,371 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack/lapack64"
+)
+
+var (
+	tridiagDense *Tridiag
+	_            Matrix           = tridiagDense
+	_            allMatrix        = tridiagDense
+	_            denseMatrix      = tridiagDense
+	_            Banded           = tridiagDense
+	_            MutableBanded    = tridiagDense
+	_            RawTridiagonaler = tridiagDense
+)
+
+// A RawTridiagonaler can return a lapack64.Tridiagonal representation of the
+// receiver. Changes to the DL, D, DU slices in lapack64.Tridiagonal will be
+// reflected in the original matrix, changes to the N field will not.
+type RawTridiagonaler interface {
+	RawTridiagonal() lapack64.Tridiagonal
+}
+
+// Tridiag represents a tridiagonal matrix by its three diagonals.
+type Tridiag struct {
+	mat lapack64.Tridiagonal
+}
+
+// NewTridiag creates a new n×n tridiagonal matrix with the first sub-diagonal
+// in dl, the main diagonal in d and the first super-diagonal in du. If dl, d,
+// and du are nil, new backing slices will be allocated for them. If dl and du
```

Comment:
This is ambiguous; it could be read as "if any of the inputs are nil, those nil inputs will be created at the appropriate size", but the behaviour is that if any of them are not nil, they must be already correctly sized. I think the behaviour is correct, but the documentation could be more clear. Adding "all of", "If all of dl, d and du..." would do this I think.

Reply:
I'm glad you're bringing up the correctness of the behavior. I thought about being completely flexible and allowing nil for any of dl, d, du but then didn't see much benefit in it and went with the current stricter version.

I'll add "all of", thanks.

Reply:
Done.